namespace.c

```diff
@@ -0,0 +1,1139 @@
+    ns->ns_object = 0;
+    ns->ns_id = 0;
+
+    ns->top_self = 0;
+    ns->load_path = rb_ary_dup(vm->load_path);
+    ns->expanded_load_path = rb_ary_dup(vm->expanded_load_path);
+    ns->load_path_snapshot = rb_ary_new();
+    ns->load_path_check_cache = 0;
+    ns->loaded_features = rb_ary_dup(vm->loaded_features);
+    ns->loaded_features_snapshot = rb_ary_new();
+    ns->loaded_features_index = st_init_numtable();
+    ns->loaded_features_realpaths = rb_hash_dup(vm->loaded_features_realpaths);
+    ns->loaded_features_realpath_map = rb_hash_dup(vm->loaded_features_realpath_map);
+    ns->loading_table = st_init_strtable();
+    ns->ruby_dln_libmap = rb_hash_new_with_size(0);
+    ns->gvar_tbl = rb_hash_new_with_size(0);
+
+    ns->is_builtin = false;
+    ns->is_user = true;
+    ns->is_optional = true;
+}
+
+void rb_namespace_gc_update_references(void *ptr)
+{
+    rb_namespace_t *ns = (rb_namespace_t *)ptr;
+    if (!NIL_P(ns->ns_object))
+        ns->ns_object = rb_gc_location(ns->ns_object);
+    ns->top_self = rb_gc_location(ns->top_self);
+    ns->load_path = rb_gc_location(ns->load_path);
+    ns->expanded_load_path = rb_gc_location(ns->expanded_load_path);
+    ns->load_path_snapshot = rb_gc_location(ns->load_path_snapshot);
+    if (ns->load_path_check_cache) {
+        ns->load_path_check_cache = rb_gc_location(ns->load_path_check_cache);
+    }
+    ns->loaded_features = rb_gc_location(ns->loaded_features);
+    ns->loaded_features_snapshot = rb_gc_location(ns->loaded_features_snapshot);
+    ns->loaded_features_realpaths = rb_gc_location(ns->loaded_features_realpaths);
+    ns->loaded_features_realpath_map = rb_gc_location(ns->loaded_features_realpath_map);
+    ns->ruby_dln_libmap = rb_gc_location(ns->ruby_dln_libmap);
+    ns->gvar_tbl = rb_gc_location(ns->gvar_tbl);
+}
+
+void
+rb_namespace_entry_mark(void *ptr)
+{
+    const rb_namespace_t *ns = (rb_namespace_t *)ptr;
+    rb_gc_mark(ns->ns_object);
+    rb_gc_mark(ns->top_self);
+    rb_gc_mark(ns->load_path);
+    rb_gc_mark(ns->expanded_load_path);
+    rb_gc_mark(ns->load_path_snapshot);
+    rb_gc_mark(ns->load_path_check_cache);
+    rb_gc_mark(ns->loaded_features);
+    rb_gc_mark(ns->loaded_features_snapshot);
+    rb_gc_mark(ns->loaded_features_realpaths);
+    rb_gc_mark(ns->loaded_features_realpath_map);
+    if (ns->loading_table) {
+        rb_mark_tbl(ns->loading_table);
+    }
+    rb_gc_mark(ns->ruby_dln_libmap);
+    rb_gc_mark(ns->gvar_tbl);
+}
+
+#define namespace_entry_free RUBY_TYPED_DEFAULT_FREE
+// TODO: free all objects in rb_namespace_t, including st_table members?
+
+static size_t
+namespace_entry_memsize(const void *ptr)
+{
+    return sizeof(rb_namespace_t);
+}
+
+const rb_data_type_t rb_namespace_data_type = {
+    "Namespace::Entry",
+    {
+        rb_namespace_entry_mark,
+        namespace_entry_free,
+        namespace_entry_memsize,
+        rb_namespace_gc_update_references,
+    },
+    0, 0, RUBY_TYPED_FREE_IMMEDIATELY
```

Comment:
I'm not totally sure about the architecture, but I assume `NameSpace::Entry` objects live for a long time.  Should we make them `RUBY_TYPED_WB_PROTECTED`?

Reply:
The lifecycle of `Namespace::Entry` is equal to `Namespace`, as long as the life of a namespace itself and will not be collected shortly. Is it better to specify `RUBY_TYPED_WB_PROTECTED` in such case?

Reply:
I fixed the type so.